stream/src/main/java/org/apache/kafka/clients/processor/internals/KStreamThread.java

```diff
@@ -0,0 +1,308 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.processor.internals;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceCallback;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.processor.PTopology;
+import org.apache.kafka.clients.processor.ProcessorConfig;
+import org.apache.kafka.clients.processor.ProcessorProperties;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.Count;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.common.utils.SystemTime;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+
+public class KStreamThread extends Thread {
+
+    private static final Logger log = LoggerFactory.getLogger(KStreamThread.class);
+
+    private final Class<? extends PTopology> topologyClass;
+    private final ArrayList<StreamGroup> streamGroups = new ArrayList<>();
+    private final Map<Integer, ProcessorContextImpl> kstreamContexts = new HashMap<>();
+    private final IngestorImpl ingestor;
+    private final RecordCollectorImpl collector;
+    private final ProcessorProperties properties;
+    private final ProcessorConfig config;
+    private final Metrics metrics;
+    private final KafkaStreamingMetrics streamingMetrics;
+    private final Time time;
+    private volatile boolean running;
+    private long lastCommit;
+    private long nextStateCleaning;
+    private long recordsProcessed;
```

Comment:
This `recordsProcessed` field is never changed. If it were, it'd probably need to be made volatile, or better yet changed to be `final AtomicLong` so that operations are atomic.